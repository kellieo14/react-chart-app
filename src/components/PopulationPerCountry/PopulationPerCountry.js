import React, {useState, useEffect} from 'react';
import VerticalBar from '../Graphs/VerticalBar/VerticalBar';

export const sortPopulation = (countries) => {
    countries.sort((a, b) => (a.population < b.population) ? 1 : -1);
}

export const getTopPopulations = (countries, labels, populations) => {
    if (countries.length > 0) {
        for (let i = 0; i < 5; i++) {
            labels.push(countries[i].name);
            populations.push(Math.round(countries[i].population / 1000000));
        }
    }
}

export const setPopulationData = (countries, labels, populations, setData) => {
    if (countries.length > 0) {
        getTopPopulations(countries, labels, populations);
        setData({
            title: 'Highest Populated Countries',
            labels: labels,
            datasets: [
            {
                label: 'Population / Million',
                backgroundColor: 'rgb(94,120,254, .1)',
                borderColor: 'rgb(0,128,214)',
                borderWidth: 1,
                hoverBackgroundColor: 'rgb(94,120,254, .6)',
                data: populations
            }
            ]
        })
    } 
}

function PopulationPerCountry(props) {
    const countries = props.countries;
    let [data, setData] = useState({});

    useEffect(() => { 
        const labels = [];
        const populations = [];
        sortPopulation(countries);
        setPopulationData(countries, labels, populations, setData)
    }, [countries]);

    return (
        <div>
            <VerticalBar data = {data}/>
        </div>
      );
}

export default PopulationPerCountry;