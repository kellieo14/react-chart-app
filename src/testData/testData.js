export const countries = [
    {
        name: 'China', 
        region: 'Asia', 
        languages: [
            { iso639_1: "zh", iso639_2: "zho", name: "Chinese", nativeName: "中文 (Zhōngwén)"}
        ], 
        population: 1377422166,
        area: 9640011,
    }, 
    {
        name: 'India', 
        region: 'Asia',
        languages: [
            {iso639_1: "hi", iso639_2: "hin", name: "Hindi", nativeName: "हिन्दी"},
            {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
        ], 
        population: 1295210000,
        area: 3287590,
    }, 
    {
        name: 'United States of America', 
        region: 'Americas', 
        languages: [
            {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
        ], 
        population: 323947000,
        area: 9629091,
    },
    {
        name: 'Nigeria', 
        region: 'Africa', 
        languages: [
            {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
        ], 
        population: 186988000,
        area: 923768,
    }, 
    {
        name: 'France', 
        region: 'Europe', 
        languages: [
            {iso639_1: "fr", iso639_2: "fra", name: "French", nativeName: "français"}
        ], 
        population: 66710000,
        area: 640679,
    }, 
    {
        name: 'Australia', 
        region: 'Oceania', 
        languages: [
            {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
        ], 
        population: 24117360,
        area: 7692024,
    }, {
        name: 'Polarland', 
        region: 'Polar', 
        languages: [
            {iso639_1: "es", iso639_2: "spa", name: "Spanish", nativeName: "Español"}
        ], 
        population: 11239004,
        area: 1284000,
    }, {
        name: 'Otherland', 
        region: 'Other', 
        languages: [
            {iso639_1: "es", iso639_2: "spa", name: "Spanish", nativeName: "Español"}
        ], 
        population: 100,
        area: 619745,
    }
];