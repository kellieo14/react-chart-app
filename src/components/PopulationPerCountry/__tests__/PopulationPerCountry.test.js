import React from 'react';
import * as reactHooks from 'react';
import Enzyme, {mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { countries } from '../../../testData/testData';
import TestRenderer from 'react-test-renderer';
import PopulationPerCountry, { sortPopulation, getTopPopulations } from '../PopulationPerCountry';
Enzyme.configure({adapter: new Adapter()});

describe('PopulationPerCountry', () => {
    let spyUseState;
    let spyUseEffect;
    let setMockData;

    beforeEach(() => {
        setMockData = jest.fn();
        spyUseState = jest.spyOn(reactHooks, 'useState')
            .mockImplementationOnce(() => [{}, setMockData]);
        spyUseEffect = jest.spyOn(reactHooks, 'useEffect');
    });


    const expectedSortedCountries = 
            [ { name: 'China',
            region: 'Asia',
            languages: [
                { iso639_1: "zh", iso639_2: "zho", name: "Chinese", nativeName: "中文 (Zhōngwén)"}
            ], 
            population: 1377422166,
            area: 9640011 },
          { name: 'India',
            region: 'Asia',
            languages: [
                {iso639_1: "hi", iso639_2: "hin", name: "Hindi", nativeName: "हिन्दी"},
                {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
            ], 
            population: 1295210000,
            area: 3287590 },
          { name: 'United States of America',
            region: 'Americas',
            languages: [
                {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
            ], 
            population: 323947000,
            area: 9629091 },
          { name: 'Nigeria',
            region: 'Africa',
            languages: [
                {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
            ], 
            population: 186988000,
            area: 923768 },
          { name: 'France',
            region: 'Europe',
            languages: [
                {iso639_1: "fr", iso639_2: "fra", name: "French", nativeName: "français"}
            ], 
            population: 66710000,
            area: 640679 },
          { name: 'Australia',
            region: 'Oceania',
            languages: [
                {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
            ], 
            population: 24117360,
            area: 7692024 },
          { name: 'Polarland',
            region: 'Polar',
            languages: [
                {iso639_1: "es", iso639_2: "spa", name: "Spanish", nativeName: "Español"}
            ], 
            population: 11239004,
            area: 1284000 },
          { name: 'Otherland',
            region: 'Other',
            languages: [
                {iso639_1: "es", iso639_2: "spa", name: "Spanish", nativeName: "Español"}
            ], 
            population: 100,
            area: 619745 } ]

    it('should sort countries by population', () => {
        sortPopulation(countries);
        expect(countries).toStrictEqual(expectedSortedCountries);
    });

    it('should call useState and useEffect', () => {
        mount(<PopulationPerCountry countries={countries} />);
        expect(spyUseEffect).toHaveBeenCalled();
        expect(spyUseState).toHaveBeenCalled();
        expect(setMockData).toHaveBeenCalled();
    });

    it('should set top 5 populated countries', () => {
        let labels = [];
        let populations = [];
        getTopPopulations(expectedSortedCountries, labels, populations);
        expect(labels).toStrictEqual(['China', 'India', 'United States of America', 'Nigeria', 'France']);
        expect(populations).toStrictEqual([1377422166, 1295210000, 323947000, 186988000, 66710000]);
    });

    it('should match snapshot', () => {
           const testRenderer = TestRenderer.create(<PopulationPerCountry />);
           expect(testRenderer.toJSON()).toMatchSnapshot();
        });

});