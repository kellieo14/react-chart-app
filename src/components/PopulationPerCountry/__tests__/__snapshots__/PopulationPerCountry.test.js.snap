// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopulationPerCountry should match snapshot 1`] = `
<div>
  <div
    className="verticalBar"
  >
    <h4
      className="chartTitle"
    />
  </div>
</div>
`;
